### PR TITLE
Add argument/status checking to methods in Channel class

### PR DESCRIPTION
**Pull request by hachi8833**
I found that the methods in `Channel` were unsafe:

- when `close` is called twice
- when `receive` is called against the closed channel
- when `deliver` is called against the closed channel

Using a status flag `ChannelState` might look unsophisticated. I'd be glad if someone else have a better idea for this.
But I think the methods are now safer than before, or it might be safer than raw Golang 😆 

Well, I'm wondering how to release the ChannelObject when `closed`.

### CODE REVIEW

File: vm/channel.go [hunk 1]
```diff
@@ -63,9 +63,22 @@ import (
...
+// Channel's state.
+// To Goby language contributors: Golang's channels should be carefully handled because:
+// - You cannot write to closed channels, or got a panic.
+// - You cannot close closed channels, or got a panic.
+// - You cannot write to nil channels, or causes a deadlock.
+// - You cannot read nil channels, or causes a deadlock.
+// Ref: https://beatsync.net/main/log20150325.html
+const (
+	OPEN   int = iota
+	CLOSED int = iota
```
Comment (st0012) [inline vm/channel.go]: This `iota` is redundant I think

File: vm/channel.go [hunk 2]
```diff
@@ -118,16 +130,26 @@ func builtinChannelInstanceMethods() []*BuiltinMethodObject {
...
 			Fn: func(receiver Object, sourceLine int) builtinMethodBody {
 				return func(t *Thread, args []Object, blockFrame *normalCallFrame) Object {
+					if len(args) != 0 {
+						return t.vm.InitErrorObject(errors.ArgumentError, sourceLine, "Expect 0 argument. got: %d", len(args))
+					}
 					c := receiver.(*ChannelObject)
 
-					close(c.Chan)
+					if c.ChannelState == CLOSED {
+						return t.vm.InitErrorObject(errors.InternalError, sourceLine, "The channel is already closed. got: %d", len(args))
```
Comment (st0012) [inline vm/channel.go]: I think we should have a dedicated error type for this

Comment (hachi8833) [reply vm/channel.go]: Got it

File: vm/channel.go (hunk 1 above)
Comment (hachi8833) [reply vm/channel.go]: I'd remove the second one

File: vm/channel.go (hunk 2 above)
Comment (hachi8833) [reply vm/channel.go]: Updated

File: vm/channel.go (hunk 1 above)
Comment (hachi8833) [reply vm/channel.go]: Updated

File: vm/channel.go [hunk 3]
```diff
@@ -63,9 +63,22 @@ import (
...
 
+// Channel's state.
+// To Goby language contributors: Golang's channels should be carefully handled because:
+// - You cannot write to closed channels, or got a panic.
+// - You cannot close closed channels, or got a panic.
+// - You cannot write to nil channels, or causes a deadlock.
+// - You cannot read nil channels, or causes a deadlock.
+// Ref: https://beatsync.net/main/log20150325.html
+const (
+	OPEN = iota
```
Comment (eliothedeman) [inline vm/channel.go]: Probably want to prefix with some identifier that these refer to channel states. Also, prefer the go convention of using camel case. These also probably shouldn't be exported values. Packages outside of vm should not be touching these state values.

Comment (hachi8833) [reply vm/channel.go]: Thank you! I missed the point.

File: vm/channel.go [hunk 4]
```diff
@@ -13,7 +14,6 @@ import (
 // `Channel#new` is available.
 //
 // Note that channels are not like files and you don't need to explicitly close them (e.g.: exiting a loop).
-// If you `close` the closed channel, Goland will cause a panic. TODO: this panic should be avoided.
```
[the remark this answers is not in the document]
Comment (hachi8833) [reply vm/channel.go]: I'll fix this

[the remark this answers is not in the document]
Comment (hachi8833) [reply vm/channel.go]: Updated the sentence

File: vm/channel.go (hunk 3 above)
Comment (hachi8833) [reply vm/channel.go]: Sorry for the delay: updated the constants

**Comment by hachi8833** [comment]
I think it's ready to merge. Could you do that?

**Pull request merged**